Linux-0.98/Yggdrasil-0.98.3/usr/src/usr.bin/Archivers/zoo/fiz.1
.\" @(#) fiz.1 1.2 88/01/31 23:22:04
.\"
.\" For formatting with nroff:
.\"   nroff -man fiz.1
.\"
.TH FIZ 1 "Jan 31, 1988"
.SH NAME
fiz \- analyze damaged zoo archive for data revovery
.SH SYNOPSIS
.I fiz
.RB archive[ .zoo ]
.SH DESCRIPTION
.I Fiz
is used to analyze damaged
.I zoo
archives and locate directory entries and file data in them.
The current version of 
.I fiz
is 2.0 and it is meant to be used in conjunction with
.I zoo
version 2.0.
.I Fiz
makes no assumptions about archive structure.  Instead, it simply
searches the entire subject archive for tag values
that mark the locations of directory entries and file data.
In a 
.I zoo
archive, a
.I directory entry
contains information about a stored file such as its name, whether 
compressed or not, and its timestamp.  The
.I file data
are the actual data for the archived file, and may be
either the original data, or the result of compressing the file.
.PP
For each directory entry found,
.I fiz
prints where in the archive it is located, the directory path and
filename(s) found in it, whether the directory entry appears
to be corrupted (indicated by [*CRC Error*]), and the value of
the pointer to the file data that is found in the directory entry.
For each block of file data found in the archive,
.I fiz
prints where in the archive the block begins.  In the case of
an undamaged archive, the pointer to file data found in
a directory entry will correspond to where
.I fiz
actually locates the data.  Here is some sample output from 
.I fiz:
.PP
.nf
****************
    2526: DIR  [changes] ==>   95
    2587: DATA
****************
    3909: DIR  [copyrite] ==> 1478
    3970: DATA
    4769: DATA
****************
.fi
.sp 1
In such output, 
.B DIR
indicates where 
.I fiz
found a directory entry in the archive, and
.B DATA
indicates where 
.I fiz
found file data in the archive.  Filenames located by
.I fiz
are enclosed in square brackets, and the notation
"==>   95" indicates that the directory entry found by
.I fiz
at position 2526 has a file data pointer to
position 95.  In actuality,
.I fiz
found file data at positions 2587, 3970, and
4769.  Since 
.I fiz
found only two directory entries, and each directory entry corresponds
to one file, one of the file data positions is an artifact.
.PP
Once the locations of directory entries and file data are found, the
.B @
modifier to
.I "zoo's"
archive list and extract commands can be used and
the archive contents selectively listed or extracted,
skipping the damaged portion.  This is further described
in the documentation for
.I zoo(1).
.PP
In the above case, commands to try giving to 
.I zoo
might be
.B x@2526,2587
(extract beginning at position 2526, and get file data
from position 2587),
.B x@3090,3970
(extract at 3090, get data from 3970)
and
.B x@3909,4769
(extract at 3909, get data from 4769).  Once a correctly-matched
directory entry/file data pair is found,
.I zoo
will in most cases synchronize with and correctly extract all files
subsequently found in the archive.  Trial and error should allow
all undamaged files to be extracted.
Also note that self-extracting archives created using
.I sez
(the Self-Extracting
.I Zoo
utility for MS-DOS), which are normally executed on an MS-DOS
system for extraction, can
be extracted on non-MSDOS systems in a similar way.
.PP
.SH "SEE ALSO"
zoo(1)
.SH BUGS
Random byte patterns can occasionally be incorrectly recognized
as tag values.  This occurs very rarely, however, and trial
and error will usually permit all undamaged data to be
extracted.
.SH DIAGNOSTICS
.I Fiz
always exits with a status code of 0.
.SH "FUTURE DIRECTIONS"
Automation of data recovery from a damaged archive is potentially
achievable.  However, since damaged archives occur only rarely,
.I fiz
as it currently stands is unlikely to change much in the
near future.
.SH AUTHOR
Rahul Dhesi


Linux-0.98/Yggdrasil-0.98.3/usr/src/usr.bin/Archivers/zoo/fiz.c
#ifndef LINT
static char sccsid[]="@(#) fiz.c 2.6 88/01/31 23:23:50";
#endif /* LINT */

/*
The contents of this file are hereby released to the public domain.

                                   -- Rahul Dhesi 1987/02/06
*/

/*
Searches for all directory entries in an archive and prints their
offsets.  Zoo 1.41 and later may then be asked to extract a specific
file by supplying the offset of the file.
*/

#include "options.h"
#include "zooio.h"
#include "various.h"
#include "zoofns.h"
#include "portable.h"         /* I/O definitions */
#include "zoo.h"

void prtctrl ();
void prtch ();

main(argc,argv)
register int argc;
register char **argv;
{
   char *zooname;          /* name of archive to be read */
   ZOOFILE zoo_file;       /* the archive being examined opened for read */
   int state;              /* to keep track of how much of tag seen */
   int inch;               /* char just read from archive */

   static char usage1[] = "Fiz 2.0 (1987/02/01) public domain Zoo archive repair utility by Rahul Dhesi\n";
   static char usage2[] = "Usage:  fiz archive[.zoo]  (\"fiz -h\" for help)\n";

#ifdef SETBUF
/* set stdout to unbuffered */
setbuf (stdout, (char *) NULL);
#endif

   if (argc < 2) {
      printf("%s%s", usage1, usage2);
      exit (1);
   }

   if (strcmp(argv[1],"-h") == 0)
      goto givehelp;

   zooname = argv[1];

   /* Add default extension if none supplied */
   {
      char *p, *q;
      p = zooname + strlen(zooname);         /* point to last char */
      while (p != zooname && *p != EXT_CH)
         --p;
      /* either found EXT_CH or reached beginning of zooname */
      if (*p != EXT_CH) {
         q = malloc(strlen(zooname) + strlen(EXT_DFLT) + 2);
         if (q == NULL) {
            printf("Fiz:  Ran out of memory.\n");
            exit(1);
         }
         strcpy(q, zooname);
         strcat(q, EXT_DFLT);
         zooname = q;
      }
   }

   zoo_file = zooopen (zooname, Z_READ);
   if (zoo_file == NOFILE) {
      printf("Fiz:  FATAL:  Could not open %s.\n", zooname);
      exit(1);
   }

#ifdef DOUBLE_SECRET
	{ void oh_well(void); oh_well(); }
#endif

#define  NOSTATE  1
#define  HDR_1   0xdc
#define  HDR_2   0xa7
#define  HDR_3   0xc4
#define  HDR_4   0xfd

#define	DAT_1   '@'
#define	DAT_2   ')'
#define	DAT_3   '#'
#define	DAT_4   '('

/* finite state machine implemented here by hand */

   state = NOSTATE;
   while ((inch = zgetc(zoo_file)) != EOF) {
      inch = inch & 0xff;
      if (state == NOSTATE) {
			if (inch == HDR_1)
				state = HDR_1;
			else if (inch == DAT_1)
				state = DAT_1;
		} else if (state == HDR_1 && inch == HDR_2)
         state = HDR_2;
      else if (state == HDR_2 && inch == HDR_3)
         state = HDR_3;
      else if (state == HDR_3 && inch == HDR_4)
         state = HDR_4;
		else if (state == DAT_1 && inch == DAT_2)
			state = DAT_2;
		else if (state == DAT_2 && inch == DAT_3)
			state = DAT_3;
		else if (state == DAT_3 && inch == DAT_4)
			state = DAT_4;
      else
         state = NOSTATE;

      if (state == HDR_4) {           						/* found archive tag */
         long save_pos;
         struct direntry direntry;
         save_pos = zootell(zoo_file);
         zooseek(zoo_file, save_pos-4L, 0);				/* back to tag pos */
         frd_dir(&direntry, zoo_file);						/* read dir entry */
         printf("****************\n");

         printf ("%8lu: DIR ", save_pos-4L);

         if (direntry.dirlen > 0) {
            printf ("[");
            prtctrl (direntry.dirname);
            printf ("]");
         }

         printf(" [");
         prtctrl (direntry.fname);
         printf ("]");

         if (direntry.namlen > 0) {
            printf (" [");
            prtctrl (direntry.lfname);
            printf ("]");
         }
			printf (" ==> %4lu", direntry.offset);
         if (direntry.dir_crc != 0)
            printf (" [*bad CRC*]");
         printf ("\n");
         fseek (zoo_file, save_pos, 0);         /* try again from there */
      } else if (state == DAT_4) {				/* file data */
         printf ("%8lu: DATA\n", zootell(zoo_file) + 1);
		}
   }
exit (0);      /* don't fall through */

givehelp:

/*
vi macros:
to add printf:
:s/^.*$/printf("&\\n");/
To remove printf:
:s/^printf("\(.*\)\\n");/\1/
*/
printf("Fiz is used to help you recover data from a damaged archive.  Fiz searches\n");
printf("the specified archive for directory entries and stored files, and prints the\n");
printf("position of each one found.  Each directory entry contains a number that\n");
printf("represents the location in the archive where the file is stored;  fiz also\n");
printf("prints this position.  All numbers printed are decimal numbers.\n\n");

printf("Use Zoo version 2.00 or higher to list or extract files in the damaged\n");
printf("archive starting at a position identified by fiz.  For example, you can\n");
printf("start extracting files from archive \"badarc.zoo\" at position 1098 with the\n");
printf("command:\n\n");

printf("     zoo x@1098 badarc\n\n");

printf("Zoo will ignore the first 1098 bytes of the damaged archive and you should be\n");
printf("able to recover the undamaged files from the rest of the archive.  You can\n");
printf("also manually specify where to look for the file data with a command like\n\n");

printf("     zoo x@1098,1153\n\n");

printf("which tells zoo to use the directory entry at position 1098, but to get the\n");
printf("actual file data from position 1153 (and not from where the directory entry\n");
printf("says the data ought to be).  See the manuals for fiz and zoo for more details.\n");

exit (0);
}

/*
prtctrl() prints a string with all unprintable characters converted
to printable form.  To avoid the program running astray trying to
print damaged data, no more than MAXPRT characters are printed.
Characters with the 8th bit set are printed preceded with ~.  Control
characters are printed preceded with ^.  Both ~ and ^ may preced
the character if a control character has the 8th bit set.
*/
#define  MAXPRT      50

void prtctrl (str)
char *str;
{
   unsigned int ch;
   int count;
   count = 0;

   while (count < MAXPRT && *str != '\0') {
      ch = (unsigned) *str;
      prtch(ch);
      str++;
      count++;
   }
}

/*
Does the actual character printing for prtctrl()
*/
void prtch(ch)
unsigned int ch;
{
   /* assumes ASCII character set */
   if (ch < ' ') {                        /* ^@ through ^_ */
      printf("^%c", ch + 0x40);
   } else if (ch == 0x7f) {               /* DEL */
      printf("^?");
   } else if (ch > 0x7f) {                /* 8th bit set */
      printf("~");                        /* .. so precede with ~ */
      prtch(ch & 0x7f);                   /* slick recursive call */
   } else
      printf("%c", ch);                   /* plain char */
}
